src/mpid/ch4/netmod/ofi/ofi_dynproc.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+#include "ofi_impl.h"
+
+#define DISCONNECT_CONTEXT_ID 0xF000
+#define INITIAL_NUM_CONN 10
+
+static int dynproc_send_disconnect(int conn_id);
+static int dynproc_accept_disconnect(int conn_id);
+static int dynproc_grow_conn_table(int new_max);
+static int dynproc_get_next_conn_id(int *conn_id_out);
+
+int MPIDI_OFI_dynproc_init(void)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_CONN_MANAGER_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_CONN_MANAGER_INIT);
+
+    MPIDI_OFI_global.conn_mgr.n_conn = 0;
+    MPIDI_OFI_global.conn_mgr.max_n_conn = 0;
+    MPIDI_OFI_global.conn_mgr.conn_table = NULL;
+    MPIDI_OFI_global.conn_mgr.n_free = 0;
+    MPIDI_OFI_global.conn_mgr.free_stack = NULL;
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_CONN_MANAGER_INIT);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_OFI_dynproc_finalize(void)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_CONN_MANAGER_DESTROY);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_CONN_MANAGER_DESTROY);
+
+    if (MPIDI_OFI_global.conn_mgr.n_conn > 0) {
+        for (int i = 0; i < MPIDI_OFI_global.conn_mgr.max_n_conn; i++) {
+            switch (MPIDI_OFI_global.conn_mgr.conn_table[i].state) {
+                case MPIDI_OFI_DYNPROC_CONNECTED_CHILD:
+                    mpi_errno = dynproc_send_disconnect(i);
+                    MPIR_ERR_CHECK(mpi_errno);
+                    break;
+                case MPIDI_OFI_DYNPROC_CONNECTED_PARENT:
+                    mpi_errno = dynproc_accept_disconnect(i);
+                    MPIR_ERR_CHECK(mpi_errno);
+                    break;
```

Comment:
One concern is whether this will cause dead-lock situations. The spawning process, I believe, will generate strict tree-like hierarchy, thus won't have cycles, and it should be OK. Alternatively, we can split the loops to do the `send_disconnect` first, then a separate loop to do the `accept_disconnect`.

Reply:
Could you explain more what is happening in this commit, and if this is a new concern or is from preexisting code? This looks mostly like refactoring, in which case it should be OK.

Reply:
> 
> 
> Could you explain more what is happening in this commit, and if this is a new concern or is from preexisting code? This looks mostly like refactoring, in which case it should be OK.

Previously for children connections, it does a blocking send -- issues the send then progress loop until it completes. For parent connection, it does non-blocking recv -- issues the recv, add the request to an array. And the progress loop waiting for the request happens separately, i.e. after all the children send completes.

The new code does both blocking send (no change) and blocking recv (change). Thus raising the question of whether we may have a potential deadlock. So it is a new concern.

I believe all the processes form a strict tree-like structure, i.e. no process can be both ancestor/descendent of the same process forming send-recv cycles. Thus all the blocking send and recv should be able to complete. Thus the concern can be dismissed.

The code is much simpler to maintain with blocking send/recv as they can be easily moved around. In the future, we probably want to allow processes to disconnect before MPI_Finalize, so the simpler function wrap paves the way.

Reply:
Can we use an intercomm collective operation? Using netmod-specific pt2pt here seems unnecessary. If we use something more generic, it should be easier to raise the abstraction level and reuse this in ch4/ucx. At least, I want it to be that easy 😄.

Reply:
The connection is not only used by a single intercomm, right?

Reply:
But yes, I think with proper reference counting, the extra disconnect messages seem unnecessary. When every communicator using the connection gets destroyed, we can simply remove the connection entry locally.

Reply:
> The connection is not only used by a single intercomm, right?

Maybe I don't understand this connection table as well as I think. Does each entry represent the joining 2 different world communicators? Or is it something else.

Either way, I think the converting these from several nonblocking ops to one-by-one blocking ops is probably OK for this PR.

Reply:
> Does each entry represent the joining 2 different world communicators?

Yes. But that initial intercomm can be used to merge an intracomm or derive additional comm that need maintain the same connection (includes avtables, process maps, etc.). So I don't think we can just tear it down at the disconnect of the original intercomm.